contracts/AINFT721.sol

```diff
@@ -0,0 +1,304 @@
+        super.burn(tokenId);
+    }
+
+    ////
+    // URI & METADATA RELATED FUNCTIONS
+    ////
+
+    function _baseURI() internal view override(ERC721) returns (string memory) {
+        return baseURI;
+    }
+
+    function tokenURI(
+        uint256 tokenId
+    ) public view override(ERC721) returns (string memory) {
+        _requireMinted(tokenId);
+        uint256 currentVersion = _tokenURICurrentVersion[tokenId];
+
+        return tokenURIByVersion(tokenId, currentVersion);
+    }
+
+    /**
+     * @dev Returns the key for the metadata storage.
+     * @return The metadata storage key.
+     */
+    function _metadataStorageKey(
+        uint256 tokenId,
+        uint256 version
+    ) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(tokenId.toString(), "AINFT delimeter", version));
+    }
+
+    /**
+     * @dev The metadata storage is a mapping of token ID to metadata.
+     * @return The metadata storage.
+     */
+    function metadataStorageByVersion(
+        uint256 tokenId,
+        uint256 version
+    ) public view returns (MetadataContainer memory) {
+    
+        bytes32 key = _metadataStorageKey(tokenId, version);
+        return _metadataStorage[key];
+    }
+
+    function tokenURICurrentVersion(
+        uint256 tokenId
+    ) public view returns (uint256) {
+        _requireMinted(tokenId);
+        return _tokenURICurrentVersion[tokenId];
+    }
+
+    function tokenURIByVersion(
+        uint256 tokenId,
+        uint256 uriVersion
+    ) public view returns (string memory) {
+        _requireMinted(tokenId);
+        if (uriVersion == 0) {
+            return string(abi.encodePacked(baseURI, tokenId.toString()));
+        } else {
+            MetadataContainer memory metadata = metadataStorageByVersion(tokenId, uriVersion);
+            return metadata.metadataURI;
+        }
+    }
+
+    function setBaseURI(
+        string memory newBaseURI
+    ) public onlyRole(DEFAULT_ADMIN_ROLE) returns (bool) {
+        require(
+            bytes(newBaseURI).length > 0,
+            "AINFT721::setBaseURI() - Empty newBaseURI"
+        );
+        require(
+            keccak256(bytes(newBaseURI)) != keccak256(bytes(baseURI)),
+            "AINFT721::setBaseURI() - Same newBaseURI as baseURI"
+        );
+
+        baseURI = newBaseURI;
+        //TODO: emit BatchMetadataUpdate(start, end). Consider after updateTokenURI() calls.
+        return true;
+    }
+
+    /**
+     * @dev version up & upload the metadata. You should call this function externally when the token is updated.
+     */
+    function updateTokenURI(
+        uint256 tokenId,
+        string memory newTokenURI
+    ) external returns (bool) {
+        require(
+            (isApprovedOrOwner(tx.origin, tokenId) ||
+             PAYMENT_PLUGIN == _msgSender()),
+            "AINFT721::updateTokenURI() - only payment contract can call this funciton."
+        );
+        _requireMinted(tokenId);
```

Comment:
If we already set `PAYMENT_PLUGIN`, then `isApprovedOrOwner` should be ignored. right?